core/toolbox/category.js

```diff
@@ -494,24 +510,44 @@ Blockly.ToolboxCategory.prototype.hasChildren = function() {
 };
 
 /**
- * Sets whether the category is visible or not. Categories are not visible if
- * they are the child of a parent who has been collapsed.
+ * Sets whether the category is visible or not.
+ * For a category to be visible its parent category must also be expanded.
  * @param {boolean} isVisible True if category should be visible.
- * @public
+ * @private
+ */
+Blockly.ToolboxCategory.prototype.setVisible_ = function(isVisible) {
+  this.htmlDiv_.style.display = isVisible ? 'block' : 'none';
+  if (this.hasChildren()) {
+    for (var i = 0, child; (child = this.contents_[i]); i++) {
+      child.setVisible_(isVisible);
+    }
+  }
+  this.isHidden_ = !isVisible;
+};
+
+/**
+ * Hide the category.
```

Comment:
Question: In the PR description you said in certain cases the developer would need to call toolbox.clearSelection() to close the flyout; is that the case when they call this hide function on the currently selected item? If so, maybe add that instruction here so they know they need to call it?

Or is it possible to detect that this is being called on the selected item and call it for them? I wasn't sure if that was a requirement for them to call the function or if you were clarifying that if they wanted the flyout to close, they would still have to do it themselves.

Reply:
So, it doesn't technically break anything but it is definitely a bit weird to have the flyout open when the category has been hidden. 

We could do something like: 
```
if (this.parentToolbox_.getSelectedItem() == this) {
   this.parentToolbox_.clearSelection();
}
```
It seemed a bit messy and wasn't sure if we wanted to leave this up to the developer. But thinking about it more I can't think of a situation where a developer wouldn't want to do this. So I think I will add back in.